src/query/RelayQueryPath.js

```diff
@@ -41,107 +59,103 @@ const typeField = RelayQuery.Field.build({
-    parent?: RelayQueryPath
-  ) {
-    if (node instanceof RelayQuery.Root) {
-      invariant(
-        !parent,
-        'RelayQueryPath: Root paths may not have a parent.'
-      );
-      this._name = node.getName();
-    } else {
-      invariant(
-        parent,
-        'RelayQueryPath: A parent is required for field paths.'
-      );
-      this._name = parent.getName();
-    }
-    this._node = node;
-    this._parent = parent;
-  }
-
-  /**
-   * Returns true if this is a root path (the node is a root node with an ID),
-   * false otherwise.
-   */
-  isRootPath(): boolean {
-    return !this._parent;
-  }
-
-  /**
-   * Gets the parent path, throwing if it does not exist. Use `!isRootPath()`
-   * to check if there is a parent.
-   */
-  getParent(): RelayQueryPath {
-    var parent = this._parent;
+const RelayQueryPath = {
+  createForID(dataID: DataID, name: string): QueryPath {
     invariant(
-      parent,
-      'RelayQueryPath.getParent(): Cannot get the parent of a root path.'
+      !RelayRecord.isClientID(dataID),
+      'RelayQueryPath.createForID: Expected dataID to be a server id, got ' +
+      '`%s`.',
+      dataID
     );
-    return parent;
-  }
+    return {
+      dataID,
+      name,
+      type: 'node',
+    };
+  },
 
-  /**
-   * Helper to get the name of the root query node.
-   */
-  getName(): string {
-    return this._name;
-  }
+  create(root: RelayQuery.Root): QueryPath {
+    if (root.getFieldName() === NODE) {
+      const identifyingArg = root.getIdentifyingArg();
+      if (identifyingArg && typeof identifyingArg.value === 'string') {
+        return {
+          dataID: identifyingArg.value,
+          name: root.getName(),
+          type: 'node',
+        };
+      }
+    }
+    return {
+      root,
+      type: 'root',
+    };
+  },
 
-  /**
-   * Gets a new path that describes how to access the given `node` via the
-   * current path. Returns a new, root path if `dataID` is provided and
-   * refetchable, otherwise returns an extension of the current path.
-   */
   getPath(
+    parent: QueryPath,
     node: RelayQuery.Field | RelayQuery.Fragment,
     dataID: DataID
-  ): RelayQueryPath {
+  ): QueryPath {
     if (RelayRecord.isClientID(dataID)) {
-      return new RelayQueryPath(node, this);
+      return {
+        node,
+        parent,
+        type: 'child',
+      };
+    } else if (parent.type === 'node' && parent.dataID === dataID) {
+      return parent;
     } else {
-      const root = RelayQuery.Root.build(
-        this.getName(),
-        RelayNodeInterface.NODE,
+      return {
         dataID,
-        [idField, typeField],
-        {
-          identifyingArgName: RelayNodeInterface.ID,
-          identifyingArgType: RelayNodeInterface.ID_TYPE,
-          isAbstract: true,
-          isDeferred: false,
-          isPlural: false,
-        },
-        NODE_TYPE
+        name: RelayQueryPath.getName(parent),
+        type: 'node',
+      };
+    }
+  },
```

Comment:
Note that the previous version created a full query object every time `getPath` was called, the new version delays that cost until the query is needed. In addition, we reuse the parent path if the ids exactly match.